feat: use `goto` for shallow routing, deprecate `push/replaceState` (#16449)
sveltejs/kit: documentation/docs/30-advanced/67-shallow-routing.md

---

> title: History state & shallow routing

Reviewer: this title is too long, we don't have any multi-line headers in the sidebar

<img width="208" height="194" alt="image" src="https://github.com/user-attachments/assets/30975f12-69d5-4a6a-9a3c-fc09c77260a6" />

```suggestion
title: Page state & shallow routing
```

---

As you navigate around a SvelteKit app, you create _history entries_. Clicking the back and forward buttons traverses through this list of entries, re-running any `load` functions and replacing page components as necessary.

Sometimes, it's useful to create history entries _without_ navigating. For example, you might want to show a modal dialog that the user can dismiss by navigating back. This is particularly valuable on mobile devices, where swipe gestures are often more natural than interacting directly with the UI. In these cases, a modal that is _not_ associated with a history entry can be a source of frustration, as a user may swipe backwards in an attempt to dismiss it and find themselves on the wrong page.

> SvelteKit makes this possible with the [`goto`]($app-navigation#goto) function, which allows you to associate state with a history entry without navigating. For example, to implement a history-driven modal:

Reviewer: Not sure about this API. Creating a new history entry with state is just a shallow navigation to the current URL. It's very weird to me to omit `shallow: true` here. If we change that we might need to adjust the order in which concepts are introduced here

```svelte
<!--- file: +page.svelte --->
<script>
	import { goto } from '$app/navigation';
	import { page } from '$app/state';
	import Modal from './Modal.svelte';

	function showModal() {
		goto(null, {
			state: { showModal: true }
		});
	}
</script>

<button onclick={showModal}>open</button>

{#if page.state.showModal}
	<Modal close={() => history.back()} />
{/if}
```

The modal can be dismissed by navigating back (unsetting `page.state.showModal`) or by interacting with it in a way that causes the `close` callback to run, which will navigate back programmatically.

> The above sets state and creates a new navigation entry. If you don't want that, use `replace: true` (i.e. `goto(null, { state, replace: true })`). Doing so replaces the current history entry and no new entry is created.

Reviewer: ````suggestion
The above sets state and creates a new navigation entry. If you don't want that, you can replace the existing navigation entry instead:

```js
goto(null, {
  state,
  replace: true
});
```
````

`state` can be accessed through the [page object]($app-state#page) as `page.state`. You can make page state type-safe by declaring an [`App.PageState`](types#PageState) interface (usually in `src/app.d.ts`).

> By default, state set through `goto` is not restored in the browser after a reload. To preserve it across full page reloads, use `goto(..., { state, persistState: true })`. Keep in mind that `page.state` is only populated after JavaScript is loaded, so there could be flickers from changing UI elements if you use this option.

Reviewer: ````suggestion
By default, state set with `goto` is not restored after a reload. To change that, use `persistState`:

```js
goto(url, {
  state,
  persistState: true
});
```

> [!NOTE] `page.state` is only populated after JavaScript loads, which can cause flickering UI. Use it advisedly.
```

## Caveats

During server-side rendering, `page.state` is always an empty object.

Shallow routing is a feature that requires JavaScript to work. Be mindful when using it and try to think of sensible fallback behavior in case JavaScript isn't available.

> If you navigate to another page via shallow routing, reloading on that route will not start the app in shallow routing mode. Instead the actual page on that URL is loaded. On the server this is unavoidable (because history state isn't available at that point), and hence it would mean too much of a UI flicker if it would change the page and enter shallow routing once JavaScript is loaded.

Reviewer: this is duplicative with the stuff above, we could streamline it. probably best to tweak the API and then come back to the docs though

Author: This is not duplicative it isn't mentioned anywhere else at the moment. `persistState` also does not change this; we don't swap out the server-loaded page for the real page once we know we're actually shallow-routing, we only persist state (as the option name says).
